KIPs/kip-37.md

```diff
@@ -0,0 +1,746 @@
+        @param _value     The amount of tokens being transferred
+        @param _data      Additional data with no specified format
+        @return           `bytes4(keccak256("onERC1155Received(address,address,uint256,uint256,bytes)"))`
+    */
+    function onERC1155Received(address _operator, address _from, uint256 _id, uint256 _value, bytes calldata _data) external returns(bytes4);
+
+    /**
+        @notice Handle the receipt of multiple ERC1155 token types.
+        @dev An ERC1155-compliant smart contract MUST call this function on the token recipient contract, at the end of a `safeBatchTransferFrom` after the balances have been updated.
+        This function MUST return `bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))` (i.e. 0xbc197c81) if it accepts the transfer(s).
+        This function MUST revert if it rejects the transfer(s).
+        Return of any other value than the prescribed keccak256 generated value MUST result in the transaction being reverted by the caller.
+        @param _operator  The address which initiated the batch transfer (i.e. msg.sender)
+        @param _from      The address which previously owned the token
+        @param _ids       An array containing ids of each token being transferred (order and length must match _values array)
+        @param _values    An array containing amounts of each token being transferred (order and length must match _ids array)
+        @param _data      Additional data with no specified format
+        @return           `bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))`
+    */
+    function onERC1155BatchReceived(address _operator, address _from, uint256[] calldata _ids, uint256[] calldata _values, bytes calldata _data) external returns(bytes4);
+}
+```
+
+### Safe Transfer Rules
+
+To be more explicit about how the standard `safeTransferFrom` and `safeBatchTransferFrom` functions MUST operate with respect to the KIP37TokenReceiver hook functions, a list of scenarios and rules follows.
+
+#### Scenarios
+
+**_Scenario#1 :_** The recipient is not a contract.
+
+- `onKIP37Received` and `onKIP37BatchReceived` MUST NOT be called on an EOA (Externally Owned Account).
+
+**_Scenario#2 :_** The transaction is not a mint/transfer of a token.
+
+- `onKIP37Received` and `onKIP37BatchReceived` MUST NOT be called outside of a mint or transfer process.
+
+**_Scenario#3 :_** The receiver does not implement the necessary `KIP37TokenReceiver` interface function(s).
+
+- The transfer MUST be reverted with the one caveat below. - If the token(s) being sent are part of a hybrid implementation of another standard, that particular standard’s rules on sending to a contract MAY now be followed instead. See “Backwards Compatibility” section.
+
+**_Scenario#4 :_** The receiver implements the necessary `KIP37TokenReceiver` interface function(s) but returns an unknown value.
```

Comment:
Should we define rules for `ERC1155TokenReceiver`?

Reply:
Or, it would better to define a common word `tokenReceiver` instead of the specific name and define it at the beginning of this section.

Reply:
I applied it.